### PR TITLE
bugfix: s3outputstream exception when session is closing

### PR DESCRIPTION
**Pull request by jorgeaguileraseqera**
when uploading the report to S3 the session can be closed (or in progress) and an exception is fired

created a test to validate executor is closed after the upload

Signed-off-by: Jorge Aguilera <jorge.aguilera@seqera.io>

### CODE REVIEW

File: plugins/nf-amazon/src/test/com/upplication/s3fs/AwsS3NioTest.groovy [hunk 1]
```diff
@@ -1156,4 +1156,21 @@ class AwsS3NioTest extends Specification implements AwsS3BaseSpec {
         local?.deleteDir()
         deleteBucket(bucketName)
     }
+
+    def 'should close executor after upload when no sesion' () {
```
Comment (swampie) [inline plugins/nf-amazon/src/test/com/upplication/s3fs/AwsS3NioTest.groovy:1160]:
```suggestion
    def 'should close executor after upload when no session' () {
```

**Comment by pditommaso** [comment]
Not sure to follow the rationale of this PR, why did the error arise in the first place?

**Comment by jorgeaguileraseqera** [comment]
The problem comes when the user selects an S3 to upload the report. 

When the TraceObserver tries to upload the HTML it uses an S3OutputStream but this found the session was closed so will not receive a notification to close the executor and the pipeline hangs forever, so the idea is to close the executor "manually"  once the stream is closed

**Comment by pditommaso** [comment]
I got a similar error when creating the execution report 

```
Sep-03 17:27:43.442 [main] WARN  nextflow.trace.ReportObserver - Failed to render execution report -- see the log file for details
java.lang.IllegalStateException: Session shutdown already initiated — Hook cannot be added: com.upplication.s3fs.S3OutputStream$$Lambda$424/0x0000000840613c40@6d4f6815
	at nextflow.Session.onShutdown(Session.groovy:921)
	at com.upplication.s3fs.S3OutputStream.getOrCreateExecutor(S3OutputStream.java:638)
	at com.upplication.s3fs.S3OutputStream.init(S3OutputStream.java:324)
	at com.upplication.s3fs.S3OutputStream.uploadBuffer(S3OutputStream.java:305)
	at com.upplication.s3fs.S3OutputStream.flush(S3OutputStream.java:268)
	at java.base/sun.nio.cs.StreamEncoder.implFlush(StreamEncoder.java:318)
	at java.base/sun.nio.cs.StreamEncoder.flush(StreamEncoder.java:153)
	at java.base/java.io.OutputStreamWriter.flush(OutputStreamWriter.java:251)
	at java.base/java.io.BufferedWriter.flush(BufferedWriter.java:257)
	at org.codehaus.groovy.runtime.IOGroovyMethods.withWriter(IOGroovyMethods.java:1135)
	at nextflow.trace.ReportObserver.renderHtml(ReportObserver.groovy:290)
	at nextflow.trace.ReportObserver.onFlowComplete(ReportObserver.groovy:140)
	at nextflow.Session.notifyFlowComplete(Session.groovy:1073)
	at nextflow.Session.shutdown0(Session.groovy:703)
	at nextflow.Session.destroy(Session.groovy:645)
	at nextflow.script.ScriptRunner.shutdown(ScriptRunner.groovy:239)
	at nextflow.script.ScriptRunner.execute(ScriptRunner.groovy:134)
	at nextflow.cli.CmdRun.run(CmdRun.groovy:354)
	at nextflow.cli.Launcher.run(Launcher.groovy:487)
	at nextflow.cli.Launcher.main(Launcher.groovy:646)
Sep-03 17:27:43.442 [main] DEBUG nextflow.trace.TimelineObserver - Workflow completed -- rendering execution timeline
```

**Comment by pditommaso** [comment]
I'm trying a different solution https://github.com/nextflow-io/nextflow/commit/c78195eccb57a3e88470c307b2b63f255167a864. Essentially to use another shutdown hool `Global.onExit` and invoke it, just before it exists

**Comment by pditommaso** [comment]
Closing this in favour of https://github.com/nextflow-io/nextflow/commit/107ea8e7266e68fa6691a924683c149385daa32f

**Pull request closed without merging**